tasks/move-granules/index.js

```diff
@@ -59,72 +74,83 @@ function getAllGranules(input, granules, regex) {
 * @param {Object} collection - configuration object defining a collection
-* of granules and their files
+*                              of granules and their files
 * @param {string} cmrFiles - array of objects that include CMR xmls uris and granuleIds
 * @param {Object} buckets - the buckets involved with the files
-* @returns {Promise} promise resolves when all files have been moved
+* @returns {Object} new granulesObject where each granules' files are updated with
+*                   the correct target buckets/paths/and s3uri filenames.
 **/
 function updateGranuleMetadata(granulesObject, collection, cmrFiles, buckets) {
-  const allFiles = [];
+  const updatedGranules = {};
+  const fileSpecs = collection.files;
   Object.keys(granulesObject).forEach((granuleId) => {
+    const updatedFiles = [];
+    updatedGranules[granuleId] = { ...granulesObject[granuleId] };
+
     granulesObject[granuleId].files.forEach((file) => {
-      collection.files.forEach((fileConfig) => {
-        const filenameWORenameSuffix = isFileRenamed(file.name)
-          ? file.name.split('.').slice(0, -1).join('.') : file.name;
-        const match = filenameWORenameSuffix.match(fileConfig.regex);
-
-        if (match) {
-          if (!file.url_path) {
-            /* eslint-disable-next-line no-param-reassign */
-            file.url_path = fileConfig.url_path || collection.url_path || '';
-          }
-          const cmrFile = cmrFiles.find((f) => f.granuleId === granuleId);
-
-          const urlPath = urlPathTemplate(file.url_path, {
-            file: file,
-            granule: granulesObject[granuleId],
-            cmrMetadata: cmrFile ? cmrFile.metadataObject : {}
-          });
-
-          if (!buckets[fileConfig.bucket]) {
-            throw new Error(`Collection config specifies a bucket key of ${fileConfig.bucket}, but the configured bucket keys are: ${Object.keys(buckets).join(', ')}`);
-          }
-
-          /* eslint-disable no-param-reassign */
-          file.bucket = buckets[fileConfig.bucket];
-          file.filepath = path.join(urlPath, file.name);
-          file.filename = `s3://${path.join(file.bucket.name, file.filepath)}`;
-          /* eslint-enable no-param-reassign */
-
-          allFiles.push(file);
+      const match = fileSpecs.filter((cf) => unversionFilename(file.name).match(cf.regex));
+      validateMatch(match, buckets, file);
+
+      const URLPathTemplate = file.url_path || match[0].url_path || collection.url_path || '';
+      const cmrFile = cmrFiles.find((f) => f.granuleId === granuleId);
+      const urlPath = urlPathTemplate(URLPathTemplate, {
+        file,
+        granule: granulesObject[granuleId],
+        cmrMetadata: cmrFile ? cmrFile.metadataObject : {}
+      });
+      const bucket = buckets[match[0].bucket];
+      const filepath = path.join(urlPath, file.name);
+
+      updatedFiles.push({
+        ...file,
+        ...{
```

Comment:
I like this witchcraft right here.

Reply:
I want more of this witchcraft so that we have more functional code.